aptos-move/block-executor/src/executor_utilities.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{errors::*, task::ExecutorTask, view::LatestView};
+use aptos_aggregator::types::{code_invariant_error, PanicError, PanicOr};
+use aptos_logger::error;
+use aptos_mvhashmap::types::ValueWithLayout;
+use aptos_types::{
+    contract_event::TransactionEvent, executable::Executable, state_store::TStateView,
+    transaction::BlockExecutableTransaction as Transaction, write_set::TransactionWrite,
+};
+use bytes::Bytes;
+use move_core_types::value::MoveTypeLayout;
+use rand::{thread_rng, Rng};
+use std::{
+    collections::{BTreeMap, HashSet},
+    sync::Arc,
+};
+
+macro_rules! groups_to_finalize {
+    ($outputs:expr, $($txn_idx:expr),*) => {{
+	let group_write_ops = $outputs.resource_group_metadata_ops($($txn_idx),*);
+
+        group_write_ops.into_iter()
+            .map(|val| (val, false))
+            .chain([()].into_iter().flat_map(|_| {
```

Comment:
can you explain how this works?😅

Reply:
https://stackoverflow.com/questions/49455885/chain-two-iterators-while-lazily-constructing-the-second-one

This optimization doesn't even matter here in a good case, but if there is some issue with preparing written resource groups for output, you'd detect them first before you'd iterate & query multi-versioned data-structure to generate the read-needing-replacement resource groups. Might help to get a more useful error more importantly than saving those queries, but might be an overkill - lmk what you think

Reply:
Nice trick, will make a note of that😄